typescript.js

```diff
@@ -2,79 +2,60 @@ module.exports = {
-        '@typescript-eslint/no-extra-semi': 'error',
-        'no-loop-func': 'off',
-        '@typescript-eslint/no-loop-func': 'error',
-        'no-redeclare': 'off',
-        '@typescript-eslint/no-redeclare': 'error',
-        'no-shadow': 'off',
-        '@typescript-eslint/no-shadow': 'error',
-        'no-unused-expressions': 'off',
-        '@typescript-eslint/no-unused-expressions': 'error',
-        'no-unused-vars': 'off',
-        '@typescript-eslint/no-unused-vars': ['error', { vars: 'all', args: 'after-used', ignoreRestSiblings: true }],
-        'no-use-before-define': 'off',
-        '@typescript-eslint/no-use-before-define': ['error', { functions: false, classes: true, variables: true }],
-        'no-useless-constructor': 'off',
-        '@typescript-eslint/no-useless-constructor': 'error',
-        quotes: 'off',
-        '@typescript-eslint/quotes': ['error', 'single', { avoidEscape: true }],
-        'no-return-await': 'off',
-        '@typescript-eslint/return-await': 'error',
-        semi: 'off',
-        '@typescript-eslint/semi': 'error',
-        'space-before-function-paren': 'off',
-        '@typescript-eslint/space-before-function-paren': ['error', {
-            anonymous: 'always',
-            named: 'never',
-            asyncArrow: 'always'
-        }],
-        'space-infix-ops': 'off',
-        '@typescript-eslint/space-infix-ops': 'error',
+
+        // None
+
+        /*
+            TypeScript rules outside of the recommended configuration
+        */
+        '@typescript-eslint/array-type': 'error',
+        '@typescript-eslint/ban-tslint-comment': 'error',
+        '@typescript-eslint/class-literal-property-style': 'error',
+        '@typescript-eslint/consistent-indexed-object-style': 'error',
+        '@typescript-eslint/consistent-type-assertions': 'error',
+        '@typescript-eslint/consistent-type-definitions': 'error',
+        '@typescript-eslint/consistent-type-imports': 'error',
+        '@typescript-eslint/explicit-function-return-type': 'error',
+        '@typescript-eslint/explicit-member-accessibility': 'error',
+        '@typescript-eslint/member-delimiter-style': 'error',
+        '@typescript-eslint/member-ordering': 'error',
+        '@typescript-eslint/method-signature-style': 'error',
+        '@typescript-eslint/no-confusing-non-null-assertion': 'error',
+        '@typescript-eslint/no-dynamic-delete': 'error',
+        '@typescript-eslint/no-extraneous-class': 'error',
+        '@typescript-eslint/no-implicit-any-catch': 'error',
+        '@typescript-eslint/no-invalid-void-type': 'error',
+        '@typescript-eslint/no-parameter-properties': 'error',
+        '@typescript-eslint/no-require-imports': 'error',
+        '@typescript-eslint/no-type-alias': 'error',
+        '@typescript-eslint/no-unnecessary-type-constraint': 'error',
+        '@typescript-eslint/no-unused-vars-experimental': 'error',
+        '@typescript-eslint/prefer-enum-initializers': 'error',
```

Comment:
[@typescript-eslint/prefer-enum-initializers](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/prefer-enum-initializers.md)

Personally, I'm not a fan of this rule. I think it's cleaner if enum values are invisible, and I don't want to have to come up with a value for all of them. 

The justification is:
>However, by implicitly defining values, enums can lead to unexpected bugs if it's modified without paying attention to the order of its items.

I can't imagine a scenario where this would matter unless you are directly referencing an enum's value, which would be pretty bad code design. Does anyone have a real-world example to change my mind?

Reply:
I've occasionally seen enums being cast to ints in C# code so I could imagine the same happening in TypeScript code, but I agree it's not the safest thing to do. One possible use case might be for serializing an enum's value to JSON.

But neither of those are particularly compelling to me so I'm ok to disable the rule unless we find a compelling example.

Reply:
Yes, the real world use case is when an enum crosses a code boundary. The common example being values passed between client and server. In these cases the values need to be preserved. However, it's more common for enums to be used only in the client where numerical values are inferred and cleanup is avoided. I use implicit enums all the time and prefer the option as well.

Reply:
@mure to add a proposal to disable this rule

Reply:
Since Carson is OOO, here's a 
**Proposal**: Disable this rule
```js
/* If an enum is crossing a code boundary (being serialized to JSON for example) then you should initialize its values. But in most other cases code shouldn't care about the values of enum items, so you can safely leave them implicit. */
'@typescript-eslint/prefer-enum-initializers': 'off'
```

Reply:
> I can't imagine a scenario where this would matter unless you are directly referencing an enum's value

The scenario it matters is when you serialise your enum values somewhere (eg to a file, or to a database).
Without explicit initialisers if you change the order (/ add new members) after the initial serialisation then your serialised data is no longer consistent and now carries a new meaning.

---

Another usecase to think of is that without an initialiser TS will set the first enum member to have value `0`.

Which means that the following code:
```ts
enum Foo {
  A,
  B,
}

function test(arg?: Foo | null) {
  console.log('is arg falsey', !arg);
}

test();          // is arg falsey true
test(undefined); // is arg falsey true
test(null);      // is arg falsey true
test(Foo.A);     // is arg falsey true
test(Foo.B);     // is arg falsey false
```

This can easily cause unintended bugs if your codebase uses loose boolean checks.

If you always use explicit initialisers it means that you can only run into this case if you explicitly set a member to a falsey value.
You can also get around this by using a rule like `@typescript-eslint/strict-boolean-expressions`.

Reply:
@rajsite vote already